File: kukulcan-api/src/main/scala/com/github/mmolimar/kukulcan/KProducer.scala
package com.github.mmolimar.kukulcan

import _root_.java.util.Properties

import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.tools.{ToolsUtils => JToolsUtils}

import scala.collection.JavaConverters._

/**
 * Factory for [[com.github.mmolimar.kukulcan.KProducer]] instances.
 *
 */
object KProducer {

  def apply[K, V](props: Properties): KProducer[K, V] = new KProducer(props)

}

/**
 * An enriched implementation of the {@code org.apache.kafka.clients.producer.KafkaProducer} class to
 * produce messages in Kafka.
 *
 * @param props Properties with the configuration.
 */
class KProducer[K, V](val props: Properties) extends KafkaProducer[K, V](props) {

  import org.apache.kafka.common.{Metric, MetricName}

  /**
   * Get all metrics registered.
   *
   * @return a {@code Map} with the all metrics registered.
   */
  def getMetrics: Map[MetricName, Metric] = {
    getMetrics(".*", ".*")
  }

  /**
   * Get all metrics registered filtered by the group and name regular expressions.
   *
   * @param groupRegex Regex to filter metrics by group name.
   * @param nameRegex  Regex to filter metrics by its name.
   * @return a {@code Map} with the all metrics registered filtered by the group and name regular expressions.
   */
  def getMetrics(groupRegex: String, nameRegex: String): Map[MetricName, Metric] = {
    metrics.asScala
      .filter(metric => metric._1.group.matches(groupRegex) && metric._1.name.matches(nameRegex))
      .toMap
  }

  /**
   * Print all metrics.
   */
  def listMetrics(): Unit = {
    listMetrics(".*", ".*")
  }

  /**
   * Print all metrics filtered by the group and name regular expressions.
   *
   * @param groupRegex Regex to filter metrics by group name.
   * @param nameRegex  Regex to filter metrics by its name.
   */
  def listMetrics(groupRegex: String, nameRegex: String): Unit = {
    JToolsUtils.printMetrics(getMetrics(groupRegex, nameRegex).asJava)
  }

}

File: kukulcan-api/src/main/java/com/github/mmolimar/kukulcan/java/KProducer.java
package com.github.mmolimar.kukulcan.java;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.common.Metric;
import org.apache.kafka.common.MetricName;
import org.apache.kafka.tools.ToolsUtils;

import java.util.Map;
import java.util.Properties;
import java.util.stream.Collectors;

/**
 * An enriched implementation of the {@code org.apache.kafka.clients.producer.KafkaProducer} class to
 * produce messages in Kafka.
 */
public class KProducer<K, V> extends KafkaProducer<K, V> {

    /**
     * @param props Properties with the configuration.
     */
    public KProducer(Properties props) {
        super(props);
    }

    /**
     * Get all metrics registered.
     *
     * @return a {@code Map[MetricName, Metric]} with the all metrics registered.
     */
    public Map<MetricName, Metric> getMetrics() {
        return getMetrics(".*");
    }

    /**
     * Get all metrics registered filtered by the group regular expressions.
     *
     * @param groupRegex Regex to filter metrics by group name.
     * @return a {@code Map[MetricName, Metric]} with the all metrics registered filtered by the group regular
     * expression.
     */
    public Map<MetricName, Metric> getMetrics(String groupRegex) {
        return getMetrics(groupRegex, ".*");
    }

    /**
     * Get all metrics registered filtered by the group and name regular expressions.
     *
     * @param groupRegex Regex to filter metrics by group name.
     * @param nameRegex  Regex to filter metrics by its name.
     * @return a {@code Map[MetricName, Metric]} with the all metrics registered filtered by the group and name
     * regular expressions.
     */
    public Map<MetricName, Metric> getMetrics(String groupRegex, String nameRegex) {
        return metrics().entrySet().stream()
                .filter(metric -> metric.getKey().group().matches(groupRegex) &&
                        metric.getKey().name().matches(nameRegex))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    /**
     * Print all metrics.
     */
    public void listMetrics() {
        listMetrics(".*");
    }

    /**
     * Print all metrics filtered by the group regular expression.
     *
     * @param groupRegex Regex to filter metrics by group name.
     */
    public void listMetrics(String groupRegex) {
        listMetrics(groupRegex, ".*");
    }

    /**
     * Print all metrics filtered by the group and name regular expressions.
     *
     * @param groupRegex Regex to filter metrics by group name.
     * @param nameRegex  Regex to filter metrics by its name.
     */
    public void listMetrics(String groupRegex, String nameRegex) {
        ToolsUtils.printMetrics(getMetrics(groupRegex, nameRegex));
    }
}
